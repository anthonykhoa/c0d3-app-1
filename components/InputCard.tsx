import React, { useState } from 'react'
import { Button } from './theme/Button'
import { MdInput } from './MdInput'

const capitalizeFirst = (str: string) => {
  return str.replace(/./, char => char.toUpperCase())
}
type Option = {
  title: string
  placeHolder?: string
  value?: string | number
}

type Btn = {
  title: string
  onClick?: Function
}

type InputCardProps = {
  values: Option[]
  buttons?: Btn[]
  capitalizeTitle?: boolean
  bgColor?: 'white' | 'none'
  title?: string
}

type ButtonsProps = {
  buttons: Btn[]
  options: Option[]
}

const Buttons: React.FC<ButtonsProps> = ({ buttons, options }) => {
  const btns = buttons.map((btn: Btn, i: number) => (
    <div style={{ display: 'inline-block', margin: 10 }} key={i}>
      <Button
        onClick={() => btn.onClick && btn.onClick(options)}
        type="success"
      >
        {btn.title}
      </Button>
    </div>
  ))

  return <>{btns}</>
}

export const InputCard: React.FC<InputCardProps> = ({
  values,
  buttons,
  capitalizeTitle = true,
  title,
  bgColor = 'white'
}) => {
  const [options, saveOptions] = useState(values)

  const handleChange = (value: string, i: number) => {
    const newOptions = [...options]
    newOptions[i].value = value
    saveOptions(newOptions)
  }

  const inputs = options.map((obj: Option, i: number) => {
    const { title, value, placeHolder } = obj
    if (obj.title === 'id') return []
    return (
      <div
        key={i}
        className="d-flex flex-column"
        style={{
          border: '3px solid rgb(84, 64, 216, .3)',
          padding: '10px',
          backgroundColor: 'rgb(84, 64, 216, .04)',
          textAlign: 'left',
          marginBottom: 10
        }}
      >
        <h5 data-testid={`h5${title}${i}`}>
          {(capitalizeTitle && capitalizeFirst(title)) || title}
        </h5>
        {title === 'description' && (
          <MdInput
            bgColor="white"
            value={(value && value + '') || ''}
            onChange={(value: string) => handleChange(value, i)}
          />
        )}
        {title !== 'description' && (
          <input
            data-testid={`input${title}`}
            style={{ border: '1px solid rgb(84, 64, 216, .3)' }}
            type="text"
            value={(value && value + '') || ''}
            onChange={e => handleChange(e.target.value, i)}
            className="form-control"
            placeholder={placeHolder || ''}
            aria-label=""
            aria-describedby="basic-addon1"
          />
        )}
      </div>
    )
  })

  return (
    <div
      style={{
        textAlign: 'center',
        padding: 10,
        backgroundColor: bgColor,
        border: '2px solid rgb(0,0,0,.3)'
      }}
    >
      {title && (
        <h2 style={{ textDecoration: 'underline', marginBottom: 20 }}>
          {title}
        </h2>
      )}
      {inputs}
      {buttons && <Buttons buttons={buttons} options={options} />}
    </div>
  )
}
