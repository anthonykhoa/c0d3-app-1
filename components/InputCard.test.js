import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { InputCard } from './InputCard'

let testBtnOnClick = ''

const mockBtn = [
  {
    title: 'Create New Challenge',
    onClick: ok => (testBtnOnClick = ok)
  }
]

let mockValues = [{ title: 'super' }]

describe('InputCard Component', () => {
  test('Input value should be nothing when value prop is not there', () => {
    const { container, queryByTestId } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    const input = queryByTestId('inputsuper')

    expect(input.value).toEqual('')
    expect(container).toMatchSnapshot()
  })

  test('Should render text onto textbox when user types', () => {
    const { container, queryByTestId } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    const input = queryByTestId('inputsuper')

    fireEvent.change(input, {
      target: { value: 'Javascript' }
    })

    expect(input.value).toContain('Javascript')
    expect(container).toMatchSnapshot()
  })

  test('Should render textbox when title is description', () => {
    mockValues = [{ title: 'description' }]
    const { container, queryByTestId } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    const input = queryByTestId('textbox')
    fireEvent.change(input, {
      target: { value: 'Javascript' }
    })

    expect(input.value).toContain('Javascript')
    expect(container).toMatchSnapshot()
  })

  test('Call function when button is clicked', () => {
    mockValues = [{ title: 'description' }]
    const { container, queryByText } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    const button = queryByText('Create New Challenge')
    fireEvent.click(button, { button: 1 })
    expect(testBtnOnClick).toEqual([{ title: 'description' }])
    expect(container).toMatchSnapshot()
  })

  test('Should output empty title when title is empty', () => {
    mockValues = [{ title: '' }]
    const { container, getByTestId } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    const input = getByTestId('h50')
    expect(input.value).toBeFalsy
    expect(container).toMatchSnapshot()
  })

  test('Should render nothing if id is a title', () => {
    mockValues = [{ title: 'id' }]
    const { container } = render(
      <InputCard values={mockValues} buttons={mockBtn} />
    )
    expect(container).toBeFalsy
    expect(container).toMatchSnapshot()
  })
})
