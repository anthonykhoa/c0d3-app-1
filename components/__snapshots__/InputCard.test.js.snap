// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputCard Component Call function when button is clicked 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 10px;"
    >
      <h5
        data-testid="h5description0"
      >
        Description
      </h5>
      <div
        style="background-color: white;"
      >
        <button
          class="btn"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </button>
        <button
          class="btn"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </button>
        <textarea
          data-testid="textbox"
          placeholder="Type something..."
          style="padding: 1rem; width: 100%;"
        />
      </div>
    </div>
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`InputCard Component Input value should be nothing when value prop is not there 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 10px;"
    >
      <h5
        data-testid="h5super0"
      >
        Super
      </h5>
      <input
        aria-describedby="basic-addon1"
        aria-label=""
        class="form-control"
        data-testid="inputsuper"
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`InputCard Component Should output empty title when title is empty 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 10px;"
    >
      <h5
        data-testid="h50"
      />
      <input
        aria-describedby="basic-addon1"
        aria-label=""
        class="form-control"
        data-testid="input"
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`InputCard Component Should render nothing if id is a title 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`InputCard Component Should render text onto textbox when user types 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 10px;"
    >
      <h5
        data-testid="h5super0"
      >
        Super
      </h5>
      <input
        aria-describedby="basic-addon1"
        aria-label=""
        class="form-control"
        data-testid="inputsuper"
        placeholder=""
        type="text"
        value="Javascript"
      />
    </div>
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;

exports[`InputCard Component Should render textbox when title is description 1`] = `
<div>
  <div
    style="text-align: center; padding: 10px; background-color: white;"
  >
    <div
      class="d-flex flex-column"
      style="padding: 10px; text-align: left; margin-bottom: 10px;"
    >
      <h5
        data-testid="h5description0"
      >
        Description
      </h5>
      <div
        style="background-color: white;"
      >
        <button
          class="btn"
          style="color: rgb(41, 41, 41);"
        >
          Write
        </button>
        <button
          class="btn"
          style="color: rgb(142, 142, 142);"
        >
          Preview
        </button>
        <textarea
          data-testid="textbox"
          placeholder="Type something..."
          style="padding: 1rem; width: 100%;"
        >
          Javascript
        </textarea>
      </div>
    </div>
    <div
      style="display: inline-block; margin: 10px;"
    >
      <button
        class="btn bg-success"
        style="color: rgb(41, 41, 41);"
      >
        Create New Challenge
      </button>
    </div>
  </div>
</div>
`;
